### PR TITLE
Populate the file panel using the event index if available.

### PR DESCRIPTION
**Pull request by poljar**
As the title says, this pull request implements pulling events containing files out of the event index and populating the file panel with them.

This intends to fix https://github.com/vector-im/riot-web/issues/4959.

A demonstration of this can be found [here](https://streamable.com/tbvqh).

The remaining issue for now is that live events aren't added to the file panel while the panel is open. If the panel is reopened the newly arrived events are added to the top instead of the bottom.

I suspect we want to fix both of those at the same time while we add the ability for riot-web, which doesn't have event indexing, to show at least events in the live timeline in the file panel.

The main issue with that seems to be that [this](https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/components/structures/FilePanel.js#L57) filter isn't smart enough to understand decrypted events that contain URLs.

### CODE REVIEW

File: src/Lifecycle.js [hunk 1]
```diff
@@ -588,8 +588,8 @@ async function startMatrixClient(startSyncing=true) {
     Mjolnir.sharedInstance().start();
 
     if (startSyncing) {
-        await MatrixClientPeg.start();
         await EventIndexPeg.init();
+        await MatrixClientPeg.start();
```
Comment (turt2live) [inline src/Lifecycle.js]: if order is dependent, a comment would be appreciated

File: src/components/structures/FilePanel.js [hunk 2]
```diff
@@ -40,42 +41,73 @@ const FilePanel = createReactClass({
...
+                        ],
+                    },
+                },
+            },
+        );
+
+        // FIXME: we shouldn't be doing this every time we change room - see comment above.
+        // TODO: Remove this stale comment? Which comment above?
+        const filterId = await client.getOrCreateFilter("FILTER_FILES_" + client.credentials.userId,
+                                                        filter);
```
Comment (turt2live) [inline src/components/structures/FilePanel.js]: this should fit on the line above without making the linter angry

File: src/components/structures/FilePanel.js [hunk 3]
```diff
@@ -40,42 +41,73 @@ const FilePanel = createReactClass({
...
+                        "types": [
+                            "m.room.message",
+                        ],
+                    },
+                },
+            },
+        );
+
+        // FIXME: we shouldn't be doing this every time we change room - see comment above.
+        // TODO: Remove this stale comment? Which comment above?
```
Comment (turt2live) [inline src/components/structures/FilePanel.js]: hmm?

File: src/indexing/EventIndex.js [hunk 4]
```diff
@@ -17,6 +17,9 @@ limitations under the License.
 import PlatformPeg from "../PlatformPeg";
 import {MatrixClientPeg} from "../MatrixClientPeg";
 
+import * as Matrix from 'matrix-js-sdk';
```
Comment (turt2live) [inline src/indexing/EventIndex.js]: In general we should be preferring to import the classes we need, like the EventTimeline below, rather than the whole js-sdk.

File: src/indexing/EventIndex.js [hunk 5]
```diff
@@ -170,7 +173,12 @@ export default class EventIndex {
             return;
         }
 
-        const e = ev.toJSON().decrypted;
+        const jsonEvent = ev.toJSON();
+
+        let e;
+        if (ev.isEncrypted()) e = jsonEvent.decrypted;
+        else e = jsonEvent;
```
Comment (turt2live) [inline src/indexing/EventIndex.js]: could shorten this to just `const e = ev.isEncrypted() ? jsonEvent.decrypted : jsonEvent;`

File: src/indexing/EventIndex.js [hunk 6]
```diff
@@ -406,4 +414,145 @@ export default class EventIndex {
...
+
+    async populateFileTimeline(timelineSet, timeline, room, limit = 10,
+                               fromEvent = null, direction = EventTimeline.BACKWARDS) {
+        const matrixEvents = await this.loadFileEvents(room, limit, fromEvent, direction);
+
+        // Add the events to the live timeline of the file panel.
+        matrixEvents.forEach(e => {
+            if (!timelineSet.eventIdToTimeline(e.getId())) {
+                timelineSet.addEventToTimeline(e, timeline,
+                                               direction == EventTimeline.BACKWARDS);
```
Comment (turt2live) [inline src/indexing/EventIndex.js]: likewise, this can go one line up.

File: src/indexing/EventIndex.js [hunk 7]
```diff
@@ -406,4 +414,145 @@ export default class EventIndex {
...
+            }
+        });
+
+        // Set the pagination token to the oldest event that we retrieved.
+        if (matrixEvents.length > 0) {
+            timeline.setPaginationToken(matrixEvents[matrixEvents.length - 1].getId(),
+                                        EventTimeline.BACKWARDS);
```
Comment (turt2live) [inline src/indexing/EventIndex.js]: one line up

File: src/indexing/EventIndex.js [hunk 8]
```diff
@@ -406,4 +414,145 @@ export default class EventIndex {
...
+        }
+    }
+
+    paginateTimelineWindow(room, timelineWindow, direction, limit) {
+        let tl;
+
+        // TODO this is from the js-sdk, this should probably be exposed to
+        // us through the js-sdk.
+        const moveWindowCap = (titmelineWindow, timeline, direction, limit) => {
+            const count = (direction == EventTimeline.BACKWARDS) ?
```
Comment (turt2live) [inline src/indexing/EventIndex.js]: we should be using triple equals for comparisons throughout

File: src/indexing/EventIndex.js [hunk 9]
```diff
@@ -406,4 +414,145 @@ export default class EventIndex {
...
+                if (excess > 0) {
+                    timelineWindow.unpaginate(3, direction != EventTimeline.BACKWARDS);
+                }
+                return true;
+            }
+
+            return false;
+        };
+
+        // TODO these private fields should be somehow exposed in the js-sdk.
```
Comment (turt2live) [inline src/indexing/EventIndex.js]: we should do this before this lands

File: src/indexing/EventIndex.js [hunk 10]
```diff
@@ -406,4 +414,145 @@ export default class EventIndex {
...
+                                                        room, limit, token, direction);
+
+            moveWindowCap(timelineWindow, timeline, direction, limit);
+            timeline.pendingPaginate = null;
+
+            return ret;
+        };
+
+        const paginationPromise = paginationMethod(timelineWindow, tl, room,
+                                                   direction, limit);
```
Comment (turt2live) [inline src/indexing/EventIndex.js]: one line up

File: src/indexing/EventIndex.js [hunk 11]
```diff
@@ -406,4 +414,145 @@ export default class EventIndex {
...
+        if (moveWindowCap(timelineWindow, tl, direction, limit)) {
+            return Promise.resolve(true);
+        }
+
+        const paginationMethod = async (timelineWindow, timeline, room, direction, limit) => {
+            const timelineSet = timelineWindow._timelineSet;
+            const token = timeline.timeline.getPaginationToken(direction);
+
+            const ret = await this.populateFileTimeline(timelineSet, timeline.timeline,
+                                                        room, limit, token, direction);
```
Comment (turt2live) [inline src/indexing/EventIndex.js]: one line up, barely.

File: src/components/structures/FilePanel.js (hunk 3 above)
Comment (poljar) [reply src/components/structures/FilePanel.js]: The TODO is from me, the FIXME was there before. 🤷

Comment (turt2live) [reply src/components/structures/FilePanel.js]: I guess just rip both out

**Comment by poljar** [comment]
Updated now. Live events handling is still missing. 

My plan is to add listeners for `Room.timeline` and `Room.decrypted` to add the live events to the loaded FilePanel.

If someone has a better idea, please shout.

**Comment by poljar** [comment]
This has been now updated to correctly handle the remaining problematic cases:
* New events that arrive while the panel is closed are correctly added when the panel is re-opened.
* New events that arrive while the panel is open.

File: src/indexing/EventIndex.js [hunk 12]
```diff
@@ -172,10 +172,7 @@ export default class EventIndex {
         }
 
         const jsonEvent = ev.toJSON();
-
-        let e;
-        if (ev.isEncrypted()) e = jsonEvent.decrypted;
-        else e = jsonEvent;
+        const e = ev.isEncrypted() ? jsonEvent.decrypted: jsonEvent;
```
Comment (turt2live) [inline src/indexing/EventIndex.js]: nit: space before `:`

File: src/indexing/EventIndex.js [hunk 13]
```diff
@@ -311,10 +308,7 @@ export default class EventIndex {
             // consume.
             const events = filteredEvents.map((ev) => {
                 const jsonEvent = ev.toJSON();
-
-                let e;
-                if (ev.isEncrypted()) e = jsonEvent.decrypted;
-                else e = jsonEvent;
+                const e = ev.isEncrypted() ? jsonEvent.decrypted: jsonEvent;
```
Comment (turt2live) [inline src/indexing/EventIndex.js]: nit: space before `:`

**Comment by poljar** [comment]
> ftr I have no idea how any of this is supposed to work, but it looks like its in the right places and does roughly the right thing. Very much assuming that bugs will show up over time and we can fix them as we see them.

Let me add a comment in that case to explain what it does.
